File: src/main/java/com/alvincezy/tinypic2/TinifyFlowable.java
package com.alvincezy.tinypic2;

import com.intellij.openapi.vfs.VirtualFile;
import com.tinify.Result;
import com.tinify.Source;
import com.tinify.Tinify;
import org.jetbrains.annotations.NotNull;

import javax.annotation.Nullable;
import java.io.IOException;

/**
 * Created by alvince on 17-7-14.
 *
 * @author alvince.zy@gmail.com
 */
public class TinifyFlowable {

    private final VirtualFile file;
    private Source source;
    private Result result;

    public TinifyFlowable(@NotNull VirtualFile file) {
        this.file = file;
    }

    public void load() throws IOException {
        source = Tinify.fromFile(file.getPath());
    }

    public boolean performTinify() {
        if (source != null) {
            try {
                result = source.result();
                return true;
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        return false;
    }

    @NotNull
    public VirtualFile file() {
        return file;
    }

    @Nullable
    public Source source() {
        return source;
    }

    @Nullable
    public Result result() {
        return result;
    }
}

File: src/main/kotlin/com/alvincezy/tinypic2/TinifyFlowable.kt
package com.alvincezy.tinypic2

import com.alvincezy.tinypic2.tinify.Verbose
import com.intellij.openapi.vfs.VirtualFile
import com.tinify.Result
import com.tinify.Source
import com.tinify.Tinify
import java.io.IOException

/**
 * Created by alvince on 17-7-14.
 *
 * @author alvince.zy@gmail.com
 * @version 1.1.2, 2018/8/31
 * @since 1.0.0
 */
class TinifyFlowable(private val file: VirtualFile) {

    var source: Source? = null
        private set

    var result: Result? = null
        private set

    private val verbose = Verbose(file)

    @Throws(IOException::class)
    fun load() {
        source = Tinify.fromFile(file.path)
    }

    fun tinify(): Boolean {
        source ?: return false

        try {
            result = source!!.result()
            result ?: return false

            verbose.complete(result!!.size())
            result!!.toFile(file.path)
        } catch (e: IOException) {
            e.printStackTrace()
            return false
        }
        return true
    }

    fun file(): VirtualFile = file

    fun verbose(): String = verbose.log()
}
